0022-circuit-block.md

```diff
@@ -0,0 +1,300 @@
+That's obviously not suitable from an API perspective.
+
+Point 1 is limiting because it means a backend-specific author can't have an annotation that lets the box use backend-specific information without risking another compiler pass mess it up.
+For example, consider a future where backends are all error corrected and defined in terms of logical qubits, but a backend wants to use some sort of "verbatim" box to allow temporary and very low-level temporary access to the physical qubits (something akin to inline `__asm__` blocks in gcc/clang-flavoured C).
+The core of the idea is that a compiler directive might be either an additive "apply this additional behaviour" marker, or it might be a subtractive "require these temporary restrictions" marker; taking point 1 or point 2 as the default is akin to choose either the former or the latter as the only type of directive permitted, respectively.
+
+Point 3 alone is not great for the same reasons as point 2; an annotation can't know all possible passes.
+That said, a limited subset of this might be possible: an annotation might mark itself as "allow me to be ignored" or "require me to be understood".
+
+> [!IMPORTANT]
+> This needs some more thought, and potentially we'd want to have a set of "predicates" that must be upheld by transpiler passes acting on the box, for example "allow only _these_ passes to inspect it" or "no special action required".
+> We'd probably need helper methods in the pass-manager and pass infrastructure to make it easier to do the right thing in these cases.
+
+We do also have a potential option in that core Qiskit can specify a small number of "special case" annotations, which all passes are required to interpret and respect.
+This gets tricky from a backwards compatibility perspective, though - if we add a new one of these special annotations in Qiskit 1.3, say, it might invalidate code that worked happily in Qiskit 1.2.
+
+
+### Serialisation and deserialisation of annotations
+
+If we allow arbitrary annotations, we reasonably need some way of constructing / deconstructing these, without sending arbitrary code to be executed over the wire - that would be a huge security hole.
+This is the problem that QPY solves for circuits.
+The general principle is that we can rely on reconstruction on the far side of a network boundary, if the receiver only runs code that it chose to install ahead of time.
+`pickle` is not suitable for this.
+
+In this model, QPY would need to be extended in a way that allows the serialisation and deserialisation of these attributes.
+Two ways of making this extension would be:
+
+* have a plugin interface to QPY, where installed packages can register additional serialisers / deserialisers for annotations.
+* add additional keyword arguments to the `dump`/`load` functions that allow a user to pass in the additional serialisers / deserialisers manually (like how is done for JSON).
```

Comment:
i'd vote for the kwargs on dump/load for custom serializers.

since the same annotations would be showing up repeatedly in the same circuit, perhaps it could cache by id like pickle.